### PR TITLE
xds: Add route to filterchain

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds the route to FilterChain.

RELEASE NOTES: none

### CODE REVIEW

File: xds/internal/xdsclient/xds.go [hunk 1]
```diff
@@ -323,8 +323,35 @@ func processNetworkFilters(filters []*v3listenerpb.Filter) ([]HTTPFilter, error)
...
+					// server-side." - A36
+					// Can specify v3 here, as will never get to this function
+					// if v2.
+					routeU, err := generateRDSUpdateFromRouteConfiguration(hcm.GetRouteConfig(), nil, false)
+					if err != nil {
+						return nil, fmt.Errorf("failed to parse inline RDS resp: %v", err)
+					}
+					filterChain.InlineRouteConfig = &routeU // Now that all this is here, scale up listener inputs in test to persist route update
+				case nil:
+					return nil, fmt.Errorf("no RouteSpecifier: +%v", hcm)
```
Comment (easwars) [inline xds/internal/xdsclient/xds.go:351]: https://github.com/grpc/proposal/pull/237 states the following:
```
If an XdsServer implementation uses RouteConfiguration or supports any HTTP
filters other than the hard-coded Router, then
`HttpConnectionManager.route_config` and `HttpConnectionManager.rds` must be
supported and RouteConfigurations must be validated. RouteConfiguration
```

So, is this correct to unconditionally expect route configuration to be present here?

File: xds/internal/xdsclient/xds.go [hunk 2]
```diff
@@ -323,8 +323,35 @@ func processNetworkFilters(filters []*v3listenerpb.Filter) ([]HTTPFilter, error)
...
+					// previous validations made for client-side usage as RDS
+					// does not distinguish between client-side and
+					// server-side." - A36
+					// Can specify v3 here, as will never get to this function
+					// if v2.
+					routeU, err := generateRDSUpdateFromRouteConfiguration(hcm.GetRouteConfig(), nil, false)
+					if err != nil {
+						return nil, fmt.Errorf("failed to parse inline RDS resp: %v", err)
+					}
+					filterChain.InlineRouteConfig = &routeU // Now that all this is here, scale up listener inputs in test to persist route update
```
Comment (easwars) [inline xds/internal/xdsclient/xds.go:349]: Nit: This comment is long enough that it can live on a line on its own. Eitherways, it is not super clear what this comment is trying to convey.

File: xds/internal/xdsclient/xds.go [hunk 3]
```diff
@@ -491,65 +518,72 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger,
...
-		// Prefer grpc_timeout_header_max, if set.
-		dur := msd.GetGrpcTimeoutHeaderMax()
-		if dur == nil {
-			dur = msd.GetMaxStreamDuration()
-		}
-		if dur != nil {
-			d := dur.AsDuration()
-			route.MaxStreamDuration = &d
+		case *v3routepb.Route_NonForwardingAction:
+			// Expected to be used on server side.
```
Comment (easwars) [inline xds/internal/xdsclient/xds.go:584]: The `action` is not persisted in the `Route` struct. Will this be done in a different PR?

**Review by easwars — CHANGES_REQUESTED** [review]

File: xds/internal/xdsclient/xds.go (hunk 2 above)
Comment (zasweq) [reply xds/internal/xdsclient/xds.go:349]: Deleted.

File: xds/internal/xdsclient/xds.go (hunk 3 above)
Comment (zasweq) [reply xds/internal/xdsclient/xds.go:584]: Oh, yeah you're right. I forgot to add it, will add a action struct. Also, I think we have a correctness issue here in regards to the default "route has invalid action". Eric's spec specifies this: "`Route.non_forwarding_action` is expected for all Routes used on server-side and
`Route.route` continues to be expected for all Routes used on client-side; a
Route with an inappropriate `action` causes RPCs matching that route to fail." Thus, it seems like we can have a route with an inappropriate action on either client or server, just the RPC's will fail. What do you think @menghanl

File: xds/internal/xdsclient/xds.go (hunk 1 above)
Comment (zasweq) [reply xds/internal/xdsclient/xds.go:351]: I was wondering about whether this was incorrect as well. I think your hunch is right, as I thought about this too. Switched to a no op.

**Review by zasweq — COMMENTED** [review]
Few notes: moving processNetworkFilters to filter_chain.go, as per Menghan's PR comment on my other PR.

File: xds/internal/xdsclient/xds.go (hunk 3 above)
Comment (menghanl) [reply xds/internal/xdsclient/xds.go:584]: Is `Route_NonForwardingAction` server side only? If so, client should NACK if it's present?

And similarly, is `Route_Route` client side only?

Comment (zasweq) [reply xds/internal/xdsclient/xds.go:584]: No nack, just says it's "expected." Eric mentions that "A route with an inappropropriate action causes RPCs matching that route to fail" (my interpretation of this means that you can have any action, but if not not forwarding action on server or route on client fail RPC's.

File: xds/internal/xdsclient/filter_chain.go [hunk 4]
```diff
@@ -391,13 +410,19 @@ func (fci *FilterChainManager) addFilterChainsForSourcePorts(srcEntry *sourcePre
...
-	httpFilters, err := processNetworkFilters(fc.GetFilters())
+// proto and stores it in our internal representation. It also persists any
+// RouteNames which need to be queried dynamically via RDS.
+func (fci *FilterChainManager) filterChainFromProto(fc *v3listenerpb.FilterChain) (*FilterChain, error) {
+	filterChain, err := processNetworkFilters(fc.GetFilters())
 	if err != nil {
 		return nil, err
 	}
-	filterChain := &FilterChain{HTTPFilters: httpFilters}
+	// These Route Names will be dynamically queried via RDS in the wrapped
```
Comment (easwars) [inline xds/internal/xdsclient/filter_chain.go:420]: Nit:
s/Route Names/route names/
s/Filter Chain/filter chain/

`FilterChain` or `RouteConfigName` could work as well. But `Filter Chain` and `Route Name` seems to make the comment stutter a bit, while also not being as helpful as they can otherwise be.

File: xds/internal/xdsclient/client.go [hunk 5]
```diff
@@ -250,8 +250,10 @@ type InboundListenerConfig struct {
 // RouteConfigUpdate contains information received in an RDS response, which is
 // of interest to the registered RDS watcher.
 type RouteConfigUpdate struct {
+	// Name is the RDS resource name for this response, which is specified in
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:253]: Used here by rds handler. Since on the server side, there can be multiple filter chains, each with a dynamic RDS specified. Thus, I coded a rds handler to handle everything in another PR I assigned you: https://github.com/grpc/grpc-go/pull/4629/files#diff-7d577e44fa077317c431482dc9284d996bbc87f53bda2d69d73c9cf1e914435eR104. Me and Menghan discussed this last week. The two options were making this a closure in rds_handler, or adding RouteName. I tried both, and RouteName was a lottttt cleaner, and also I think it might be used in other situations going forward.

File: xds/internal/xdsclient/client.go [hunk 6]
```diff
@@ -250,8 +250,10 @@ type InboundListenerConfig struct {
...
+	// LDS Response (RouteConfigName).
+	RouteName    string
 	VirtualHosts []*VirtualHost
-
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:254]: Added back. Noted.

File: xds/internal/xdsclient/filter_chain.go (hunk 4 above)
Comment (zasweq) [reply xds/internal/xdsclient/filter_chain.go:420]: Switched.

File: xds/internal/xdsclient/client.go [hunk 7]
```diff
@@ -294,6 +296,25 @@ type HashPolicy struct {
...
+	// will fail according to A36.
+	RouteActionRoute RouteAction = iota
+	// RouteActionNonForwardingAction is the expected route type on the server
+	// side. NonForwardingAction represents when a route will generate a
+	// response directly, without forwarding to an upstream host.
+	RouteActionNonForwardingAction
+	// RouteActionUnsupported are routing types currently unsupported by grpc.
+	// According to A36, "A Route with an inappropriate action causes RPCs
+	// matching that route to fail."
+	RouteActionUnsupported
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:315]: I originally had this as iota, but switched it back because of test compatibility (i.e. existing tests wouldn't have to change anything). Looking around at client.go, to keep it consistent with the other enums/default value desired, I'll switch it back.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments!

File: xds/internal/xdsclient/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/xdsclient/client.go:253]: Why is the closure option not clean? I find adding the `RouteName` here to be not clean because it is not needed on the client-side.

[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:253]: First, the closure needed to close on a loop variable so the memory that it would refer to was always the last loop variable, and it added a lot of weirdness to. Second, it just made this part of my code: https://github.com/grpc/grpc-go/pull/4629/files#diff-7d577e44fa077317c431482dc9284d996bbc87f53bda2d69d73c9cf1e914435eR70 look really ugly, and I also think it makes sense to add the name to the update. It feels like what happens with ClusterName (even though, you're right this is only on client side, not both), and it just feels like a consequential thing to persist.

[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:253]: Also, I talked to Menghan about this, and he mentioned he was indifferent and I could choose.

[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/xdsclient/client.go:253]: Doesn't something like this work?

```golang
for routeName := range routeNamesToWatch {
  if _, ok := rh.rdsCancels[routeName]; !ok {
    func(routeName string) {
      rh.rdsCancels[routeName] = rh.parent.xdsC.WatchRouteConfig(routeName, func(update xdsclient.RouteConfigUpdate, err error) {
        rh.handleRouteUpdate(routeName, update, err)
      }
    }(routeName)
  }
}

func (rh *rdsHandler) handleRouteUpdate(routeName string, update xdsclient.RouteConfigUpdate, err error) {
}
```

[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/xdsclient/client.go:253]: We *could* have done without putting the `ClusterName` in the `ClusterUpdate` as well. And now it is part of so many tests and it is hard to fix. Without a good enough reason to have the `RouteName` in the `RouteConfigUpdate`, I prefer that we don't put it in there.

File: xds/internal/xdsclient/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/xdsclient/client.go]: Yes, let's do that.

**Review by easwars — CHANGES_REQUESTED** [review]
I think other than using the closure and fixing the enum, everything looks good. Thanks.

File: xds/internal/xdsclient/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go]: IMG_7931
The most recent commit definitely switched these around already :). This comment is on a previous commit.

File: xds/internal/xdsclient/client.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:253]: IIRC, there was some usage of the clustername, but I forgot what it was. Interesting, so RouteName is essentially going to be a maintenance problem moving forward as it touches so many fields. So I trust your code works, and if it does, I think I see what I was missing. I had essentially exactly what you had now, expect rather than have an anonymous function, I tried both a. the routeName variable, and b. a local variable. What I suspect happened is the memory did not evaluate to the desired memory, as that memory would represent the last iteration. The anonymous function calls it with routeName, which means the routeName variable being closed on will be the desired one. Will work on merging this with your rds handler change tomorrow.

[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:253]: That doesn't work, same problem as I had when I did it before. It doesn't have the correct RouteName.

[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/client.go:253]: Got rds handler to work with closure. Deleting this now.

File: xds/internal/xdsclient/v2/rds_test.go [hunk 8]
```diff
@@ -111,11 +111,11 @@ func (s) TestRDSHandleResponseWithRouting(t *testing.T) {
 					VirtualHosts: []*xdsclient.VirtualHost{
 						{
 							Domains: []string{uninterestingDomain},
-							Routes:  []*xdsclient.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsclient.WeightedCluster{uninterestingClusterName: {Weight: 1}}}},
+							Routes:  []*xdsclient.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsclient.WeightedCluster{uninterestingClusterName: {Weight: 1}}, RouteAction: xdsclient.RouteActionRoute}},
```
Comment (easwars) [inline xds/internal/xdsclient/v2/rds_test.go:114]: Optional: The `Route` now has enough fields set that it can be broken into multiple lines. This and other places in this file.

File: xds/internal/xdsclient/rds_test.go [hunk 9]
```diff
@@ -714,11 +719,11 @@ func (s) TestUnmarshalRouteConfig(t *testing.T) {
 					VirtualHosts: []*VirtualHost{
 						{
 							Domains: []string{uninterestingDomain},
-							Routes:  []*Route{{Prefix: newStringP(""), WeightedClusters: map[string]WeightedCluster{uninterestingClusterName: {Weight: 1}}}},
+							Routes:  []*Route{{Prefix: newStringP(""), WeightedClusters: map[string]WeightedCluster{uninterestingClusterName: {Weight: 1}}, RouteAction: RouteActionRoute}},
```
Comment (easwars) [inline xds/internal/xdsclient/rds_test.go:722]: Optional: Same comment as in `v2/rds_test.go`.

File: xds/internal/xdsclient/filter_chain_test.go [hunk 10]
```diff
@@ -417,6 +437,323 @@ func TestNewFilterChainImpl_Failure_BadSecurityConfig(t *testing.T) {
...
+				},
+			},
+			wantErr: "unsupported type",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, err := NewFilterChainManager(test.lis)
+			if err == nil || !strings.Contains(err.Error(), test.wantErr) {
```
Comment (easwars) [inline xds/internal/xdsclient/filter_chain_test.go:750]: Nit: Combine the two lines?

**Review by easwars — APPROVED** [review]

File: xds/internal/xdsclient/filter_chain_test.go (hunk 10 above)
Comment (zasweq) [reply xds/internal/xdsclient/filter_chain_test.go:750]: Just saying, I'm pretty sure your nitpicking your own code, as I don't remember writing this hahaha :P. Done.

File: xds/internal/xdsclient/v2/rds_test.go (hunk 8 above)
Comment (zasweq) [reply xds/internal/xdsclient/v2/rds_test.go:114]: Done.

File: xds/internal/xdsclient/rds_test.go (hunk 9 above)
Comment (zasweq) [reply xds/internal/xdsclient/rds_test.go:722]: Done.

**Review by zasweq — COMMENTED** [review]
Thanks for the comments :)!

File: xds/internal/xdsclient/filter_chain_test.go (hunk 10 above)
Comment (easwars) [reply xds/internal/xdsclient/filter_chain_test.go:749]: Got to iterate my friend :)

Comment (dfawley) [reply xds/internal/xdsclient/filter_chain_test.go:749]: I constantly nitpick my own code!  I think of it more as refining though...
